apps/arweave/src/ar_http_iface_middleware.erl

```diff
@@ -2509,6 +2572,118 @@ post_block(enqueue_block, {B, Peer}, Req, ReceiveTimestamp) ->
 	end,
 	{200, #{}, <<"OK">>, Req}.
 
+handle_post_partial_solution(Req, Pid) ->
+	{ok, Config} = application:get_env(arweave, config),
+	CMExitNode = Config#config.coordinated_mining == true andalso
+			Config#config.cm_exit_peer == not_set andalso Config#config.pool_client == true,
+	case {Config#config.pool_server, CMExitNode} of
+		{false, false} ->
+			{501, #{}, jiffy:encode(#{ error => configuration }), Req};
+		{true, _} ->
+			case check_internal_api_secret(Req) of
+				{reject, {Status, Headers, Body}} ->
+					{Status, Headers, Body, Req};
+				pass ->
+					handle_post_partial_solution_pool_server(Req, Pid)
+			end;
+		{_, true} ->
+			case check_cm_api_secret(Req) of
+				{reject, {Status, Headers, Body}} ->
+					{Status, Headers, Body, Req};
+				pass ->
+					handle_post_partial_solution_cm_exit_peer_pool_client(Req, Pid)
+			end
+	end.
+
+handle_post_partial_solution_pool_server(Req, Pid) ->
+	case read_complete_body(Req, Pid) of
+		{ok, Body, Req2} ->
+			case catch ar_serialize:json_map_to_partial_solution(
+					jiffy:decode(Body, [return_maps])) of
+				{'EXIT', _} ->
+					{400, #{}, jiffy:encode(#{ error => invalid_json }), Req2};
+				Solution ->
+					Response = ar_pool:process_partial_solution(Solution),
+					JSON = ar_serialize:partial_solution_response_to_json_struct(Response),
+					{200, #{}, ar_serialize:jsonify(JSON), Req2}
+			end;
+		{error, body_size_too_large} ->
+			{413, #{}, <<"Payload too large">>, Req};
+		{error, timeout} ->
+			{500, #{}, <<"Handler timeout">>, Req}
+	end.
+
+handle_post_partial_solution_cm_exit_peer_pool_client(Req, Pid) ->
+	case read_complete_body(Req, Pid) of
+		{ok, Body, Req2} ->
+			ar_pool:post_partial_solution(Body),
+			{200, #{}, jiffy:encode(#{}), Req2};
+		{error, body_size_too_large} ->
+			{413, #{}, <<"Payload too large">>, Req};
+		{error, timeout} ->
+			{500, #{}, <<"Handler timeout">>, Req}
+	end.
+
+handle_get_jobs(PrevOutput, Req) ->
+	{ok, Config} = application:get_env(arweave, config),
+	CMExitNode = Config#config.coordinated_mining == true andalso
```

Comment:
Can this check be refactored somewhere and deuplicated? Maybe in `ar_config.erl` or `ar_coordination.erl`?

Reply:
Added ar_coordination:is_exit_peer